src/components/Recorder/index.test.js

```diff
@@ -1,79 +1,178 @@
-/**
- * @jest-environment jsdom
- */
-import React from 'react';
-import ReactDOM from 'react-dom';
-import renderer from 'react-test-renderer';
-import { render, screen, logRoles, prettyDOM } from '@testing-library/react'
-import userEvent from '@testing-library/user-event'
-import '@testing-library/jest-dom/extend-expect'
+import React, {useState} from 'react';
+import {fireEvent, render, screen} from '@testing-library/react'
 import Recorder from './index'
-import Recording from '../Recording/index'
-import useMediaRecorder from '../../hooks/useMediaRecorder'
+import useMediaRecorder from "../../hooks/useMediaRecorder";
+
 jest.mock('../../hooks/useMediaRecorder')
 
-const stream = {
-    key: "lsdkjflds",
-    stream: "sdkfjsdf08sdf",
-    name: "2021-06-18 07:37:46"
+/**
+ * Following the Object mother pattern we have this small fn that generates a valid object
+ * that matches the structure of a recording
+ *
+ * @param {{ idNumber?: number, name?: string }} options
+ * @returns {{ stream: string, name: string, id: string }}
+ */
+function createMockRecording({
+  idNumber = Math.floor(Math.random() * 100),
+  name = new Date().toISOString().split('.')[0].split('T').join(' ')
+}) {
+  return {
+    stream: "audioUrl",
+    name,
+    id: `id${idNumber}`
+  }
 }
 
-
-beforeEach(() => {
-    // Clear all instances and calls to constructor and all methods:
-    // Recorder.mockClear();
-});
+/**
+ * Applying the same pattern above we wrap the previous recording generator
+ * to create a random list of recordings
+ *
+ * @param {number} length
+ * @returns {{ stream: string, name: string, id: string }[]}
+ */
+function createMockRecordingList(length = 10) {
+  const emptyList = new Array(length).fill(null);
+  return emptyList.map(() => createMockRecording({}))
+}
 
 /**
- * We need to mock the getUserMedia function
- * https://github.com/goldingdamien/get-user-media-mock
- * https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
+ * We create a mocked version of our hook that will interact with the component in the same exact way
+ * and will expose the same API too.
```

Comment:
This is interesting and I understand the need for this, but then how do we test the _actual_ hook, or do we no longer need to since it's functionality is now an API (and either works or doesn't)?

Reply:
To test your hook, you have two options:
### 1. Create a unit test only for your hook
There are tools for this like [`react hooks testing library`](https://react-hooks-testing-library.com/)
though you could also just create a mock component that uses your the hook inside and test the integration with it.

### 2. Write an end-to-end test that will execute your code on an actual browser where the `MediaRecorder` api will be available.

Reply:
I like the end-to-end solution, to be honest. It's the most complete and real way of testing, although it can be a pain to automate. Thanks Manu.